src/api.ts

```diff
@@ -260,6 +281,92 @@ function createApi({
   }
 
   const api: Api = {
+    async importMBTiles(filePath: string) {
+      const filePathWithExtension =
+        path.extname(filePath) === '.mbtiles' ? filePath : filePath + '.mbtiles'
+
+      let mbTilesDb: DatabaseInstance
+
+      try {
+        mbTilesDb = new Database(filePathWithExtension, {
+          // Ideally would set `readOnly` to `true` here but causes `fileMustExist` to be ignored:
+          // https://github.com/JoshuaWise/better-sqlite3/blob/230ea65ed0d7566e32d41c3d13a90fb32ccdbee6/docs/api.md#new-databasepath-options
+          fileMustExist: true,
+        })
+      } catch (_err) {
+        throw new MBTilesImportTargetMissingError(filePath)
+      }
+
+      const tilejson = mbTilesToTileJSON(mbTilesDb)
+
+      // TODO: Should this be handled in extractMBTilesMetadata?
+      if (!(tilejson.format && isValidMBTilesFormat(tilejson.format))) {
+        throw new UnsupportedMBTilesFormatError()
+      }
+
+      if (!validateTileJSON(tilejson)) {
+        mbTilesDb.close()
+        throw new MBTilesInvalidMetadataError()
+      }
+
+      const insertTileData = db.prepare<{
+        data: Buffer
+        tileHash: string
+        tilesetId: string
+      }>(
+        'INSERT INTO TileData (tileHash, data, tilesetId) VALUES (:tileHash, :data, :tilesetId)'
+      )
+
+      const insertTile = db.prepare<{
+        quadKey: string
+        tileHash: string
+        tilesetId: string
+      }>(
+        'INSERT INTO Tile (quadKey, tileHash, tilesetId) VALUES (:quadKey, :tileHash, :tilesetId)'
+      )
+
+      const tilesetId = getTilesetId(tilejson)
+
+      // TODO: Look into whether we need to have a transaction for this
+      // Concerns regarding memory due to tile data size
+      const tilesImportTransaction = db.transaction(async () => {
+        const getMBTilesQuery = mbTilesDb.prepare(
+          'SELECT zoom_level as z, tile_column as y, tile_row as x, tile_data as data FROM tiles'
+        )
+
+        // TODO: See comment about transaction above. This would be the area of concern.
+        for (const { data, x, y, z } of getMBTilesQuery.iterate() as Iterable<{
+          data: Buffer
+          x: number
+          y: number
+          z: number
+        }>) {
+          const quadKey = tileToQuadKey({ zoom: z, x, y })
+
+          const tileHash = hash(data).toString('hex')
+
+          insertTileData.run({
+            tileHash,
+            data,
+            tilesetId,
+          })
+
+          insertTile.run({
+            quadKey,
+            tileHash,
+            tilesetId,
+          })
```

Comment:
It could make sense to put these two statements in a transaction.

Reply:
these are part of a transaction already

Reply:
yeah but the massive transaction that is everything from an mbtiles import. I'm suggesting smaller transactions of just these two inserts.

Reply:
ah got it. thanks for clarifying!

Reply:
addressed in https://github.com/digidem/mapeo-mapserver/pull/19/commits/a0b7576da13a9aadb975ee9d94398c320c1dfc7c